src/worker_pool/base.ts

```diff
@@ -0,0 +1,140 @@
+import assert from 'node:assert';
+
+export abstract class AsynchronouslyCreatedResource {
+    onreadyListeners : (() => void)[] | null = [];
+    ondestroyListeners : (() => void)[] | null = [];
+
+    markAsReady () : void {
+      const listeners = this.onreadyListeners;
+      assert(listeners !== null);
+      this.onreadyListeners = null;
+      for (const listener of listeners) {
+        listener();
+      }
+    }
+
+    isReady () : boolean {
+      return this.onreadyListeners === null;
+    }
+
+    onReady (fn : () => void) {
+      if (this.onreadyListeners === null) {
+        fn(); // Zalgo is okay here.
+        return;
+      }
+      this.onreadyListeners.push(fn);
+    }
+
+    onDestroy (fn : () => void) {
+      if (this.ondestroyListeners === null) {
+        return;
+      }
+
+      this.ondestroyListeners.push(fn);
+    }
+
+    markAsDestroyed () {
+      const listeners = this.ondestroyListeners;
+      assert(listeners !== null);
+      this.ondestroyListeners = null;
+      for (const listener of listeners) {
+        listener();
+      }
+    }
+
+    isDestroyed () {
+      return this.ondestroyListeners === null;
+    }
+
+    abstract currentUsage() : number;
+}
+
+export class AsynchronouslyCreatedResourcePool<
+  T extends AsynchronouslyCreatedResource> {
+  pendingItems = new Set<T>();
+  readyItems = new Set<T>();
+  maximumUsage : number;
+  onAvailableListeners : ((item : T) => void)[];
+  onTaskDoneListeners : ((item : T) => void)[];
+
+  constructor (maximumUsage : number) {
+    this.maximumUsage = maximumUsage;
+    this.onAvailableListeners = [];
+    this.onTaskDoneListeners = [];
+  }
+
+  add (item : T) {
+    this.pendingItems.add(item);
+    item.onReady(() => {
+      /* istanbul ignore else */
+      if (this.pendingItems.has(item)) {
+        this.pendingItems.delete(item);
+        this.readyItems.add(item);
+        this.maybeAvailable(item);
+      }
+    });
+  }
+
+  delete (item : T) {
+    this.pendingItems.delete(item);
+    this.readyItems.delete(item);
+  }
+
+  findAvailable () : T | null {
+    let minUsage = this.maximumUsage;
+    let candidate = null;
+    for (const item of this.readyItems) {
+      const usage = item.currentUsage();
+      if (usage === 0) return item;
+      if (usage < minUsage) {
+        candidate = item;
+        minUsage = usage;
+      }
+    }
+
+    return candidate;
+  }
```

Comment:
That code should be moved to the default load balancer function and use the workers args: worker load balancing is meant to be totally untangled from rest.

Reply:
Good catch, this is a leftover mostly; I'll clean this up

Reply:
As learning examples for devs: 
- the reimplementation of the current LB using the new API.
- And also a round robbin one: very low overhead but still efficient under load.